src/main/java/core/basesyntax/ArrayList.java

```diff
@@ -1,48 +1,164 @@
 package core.basesyntax;
 
+import java.util.Arrays;
+import java.util.NoSuchElementException;
+import java.util.Objects;
+
 public class ArrayList<T> implements List<T> {
-    @Override
-    public void add(T value) {
 
+    private static final int DEFAULT_CAPACITY = 10;
+    private static final double INCREMENT_VALUE = 1.5;
+    private int currentCapacity = DEFAULT_CAPACITY;
+    private transient Object[] elementData = new Object[DEFAULT_CAPACITY];
+    private int size;
+
+    private Object[] grow(int minCapacity) {
+        int oldCapacity = elementData.length;
+        if (oldCapacity * INCREMENT_VALUE > minCapacity) {
+            currentCapacity = (int) (oldCapacity * INCREMENT_VALUE);
+            return elementData = Arrays.copyOf(elementData, currentCapacity);
+        } else {
+            return elementData = Arrays.copyOf(elementData, (int) (minCapacity * INCREMENT_VALUE));
+        }
+    }
+
+    private Object[] grow() {
+        return grow(size + 1);
+    }
+
+    private void rangeCheckForAdd(int index) {
+        if (index > size || index < 0) {
+            throw new ArrayListIndexOutOfBoundsException("Index is incorrect " + index);
+        }
+    }
```

Comment:
Consider refactoring the index check logic into a separate method to avoid code repetition in `get`, `set`, and `remove` methods. This aligns with the DRY principle mentioned in the checklist.